File: solutions/typescript/binary-search/1/binary-search.ts
function isSorted(arr: number[]): boolean {
  return arr.every((v, i, a) => {
    if (a[i + 1]) {
      if (v > a[i + 1]) {
        return false;
      }
    }
    return true;
  });
}

export default class BinarySearch {
  array: number[] | undefined;

  constructor(array: number[]) {
    this.array = isSorted(array) ? array : undefined;
  }

  indexOf(value: number): number {
    if (this.array === undefined) {
      return -1;
    }

    let left = 0;
    let right = this.array.length - 1;

    let curr = Math.floor((left + right) / 2);

    while (left <= right) {
      const v = this.array[curr];

      if (v === value) return curr;

      if (v < value) {
        left = curr + 1;
      }

      if (v > value) {
        right = curr - 1;
      }

      curr = Math.floor((left + right) / 2);
    }

    return -1;
  }
}

File: solutions/csharp/binary-search/1/BinarySearch.cs
public static class BinarySearch
{
    public static int Find(int[] input, int value)
    {
        var left = 0;
        var right = input.Length - 1;

        while (left <= right)
        {
            var mid = (left + right) / 2;

            if (value == input[mid])
            {
                return mid;
            }

            if (value < input[mid])
            {
                right = mid - 1;
            }
            else
            {
                left = mid + 1;
            }
        }

        return -1;
    }
}